dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,695 @@
+            self.startTime = startTime
+            self.endTime = endTime
+            self.timeBound = timeBound
+            self.numMinted = 0 as UInt64
+
+            emit CollectionGroupCreated(
+                id: self.id,
+                name: self.name,
+                description: self.description,
+                typeName: self.typeName,
+                startTime: self.startTime,
+                endTime: self.endTime,
+                timeBound: self.timeBound
+            )
+        }
+    }
+
+    // Get the publicly available data for a CollectionGroup by id
+    //
+    pub fun getCollectionGroupData(id: UInt64): DSSCollection.CollectionGroupData {
+        pre {
+            DSSCollection.collectionGroupByID[id] != nil: "Cannot borrow collection group, no such id"
+        }
+
+        return DSSCollection.CollectionGroupData(id: id)
+    }
+
+    // Get the publicly available data for a Slot by id
+    //
+    pub fun getSlotData(id: UInt64): DSSCollection.SlotData {
+        pre {
+            DSSCollection.slotByID[id] != nil: "Cannot borrow slot, no such id"
+        }
+
+        return DSSCollection.SlotData(id: id)
+    }
+
+    // Get the publicly available data for a Slot by id
+    //
+    pub fun getItem(id: UInt64): DSSCollection.Item {
+        return DSSCollection.itemByID[id]!
+    }
+
+    // Validate time range of collection group
+    //
+    pub fun validateTimeRange(timeBound: Bool, startTime: UFix64?, endTime: UFix64?): Bool {
+        if !timeBound {
+            return true
+        }
+
+        if startTime! <= getCurrentBlock().timestamp && endTime! >= getCurrentBlock().timestamp {
+            return true
+        } else {
+            return false
+        }
+    }
+
+    // A DSSCollection NFT
+    //
+    pub resource NFT: NonFungibleToken.INFT, MetadataViews.Resolver {
+        pub let id: UInt64
+        pub let collectionGroupID: UInt64
+        pub let serialNumber: UInt64
+        pub let completionDate: UFix64
+        pub let completedBy: String
+        pub let level: UInt8
+
+        pub fun name(): String {
+            let collectionGroupData: DSSCollection.CollectionGroupData
+                = DSSCollection.getCollectionGroupData(id: self.collectionGroupID)
+            let level: String = self.level.toString()
+            return collectionGroupData.name
+                .concat(" Level ")
+                .concat(level)
+                .concat(" Completion Token")
+        }
+
+        pub fun description(): String {
+            let serialNumber: String = self.serialNumber.toString()
+            let completionDate: String = self.completionDate.toString()
+            return "Completed by "
+                .concat(self.completedBy)
+                .concat(" on ")
+                .concat(completionDate)
+                .concat(" with serial number ")
+                .concat(serialNumber)
+        }
+
+        destroy() {
+            emit CollectionNFTBurned(id: self.id)
```

Comment:
i think we should also decrement `totalSupply` here. Since we have `numMinted`

Reply:
Done. Thank you.